File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/PDFDocument.java
package com.jcons.pdfconverter.pdfconverter;


public class PDFDocument extends Base {

    private Header mHeader;
    private Body mBody;
    private CrossReferenceTable mCRT;
    private Trailer mTrailer;

    public PDFDocument() {
        mHeader = new Header();
        mBody = new Body();
        mBody.setByteOffsetStart(mHeader.getPDFStringSize());
        mBody.setObjectNumberStart(0);
        mCRT = new CrossReferenceTable();
        mTrailer = new Trailer();
        setId(Indentifiers.generateId());
    }

    public IndirectObject newIndirectObject() {
        return mBody.getNewIndirectObject();
    }

    public void includeIndirectObject(IndirectObject iobj) {
        mBody.includeIndirectObject(iobj);
    }

    public void setId(String id) {
        mTrailer.setId(id);
    }

    @Override
    public String toPDFString() {
        StringBuilder sb = new StringBuilder();
        sb.append(mHeader.toPDFString());
        sb.append(mBody.toPDFString());
        mCRT.setObjectNumberStart(mBody.getObjectNumberStart());
        int x = 0;
        while (x < mBody.getObjectsCount()) {
            IndirectObject iobj = mBody.getObjectByNumberID(++x);
            if (iobj != null) {
                mCRT.addObjectXRefInfo(iobj.getByteOffset(), iobj.getGeneration(), iobj.getInUse());
            }
        }
        mTrailer.setObjectsCount(mBody.getObjectsCount());
        mTrailer.setCrossReferenceTableByteOffset(sb.length());
        return sb.toString() + mCRT.toPDFString() + mTrailer.toPDFString();
    }

    @Override
    public void clear() {
        mHeader.clear();
        mBody.clear();
        mCRT.clear();
        mTrailer.clear();
    }
}

File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/PDFDocument.kt
package com.jcons.pdfconverter.pdfconverter


class PDFDocument : Base {

    private val mHeader: Header = Header()
    private val mBody: Body = Body()
    private val mTrailer: Trailer = Trailer()

    fun newIndirectObject(): IndirectObject {
        return mBody.newIndirectObject
    }

    fun includeIndirectObject(iobj: IndirectObject?) {
        mBody.includeIndirectObject(iobj!!)
    }

    fun setId(id: String?) {
        mTrailer.setId(id)
    }

    override fun toPDFString(): String {
        val sb = StringBuilder()
        sb.append(mHeader.toPDFString())
        sb.append(mBody.toPDFString())
        mTrailer.setObjectsCount(mBody.objectsCount)
        mTrailer.setCrossReferenceTableByteOffset(sb.length)
        return sb.toString() + mTrailer.toPDFString()
    }

    override fun clear() {
        mHeader.clear()
        mBody.clear()
        mTrailer.clear()
    }

    init {
        mBody.setByteOffsetStart(mHeader.pDFStringSize)
        mBody.objectNumberStart = 0
        setId(Indentifiers.generateId())
    }
}
